Draft design doc for restoring API group version by priority level (#3050)
vmware-tanzu/velero: design/restore-with-EnableAPIGroupVersions-feature.md

# Restore API Group Version by Priority Level When EnableAPIGroupVersions Feature is Set

Status: Draft

## Abstract

> This document proposes a solution to select an API group version to restore from the versions backed up using the feature EnableAPIGroupVersions.

Reviewer: ```suggestion
This document proposes a solution to select an API group version to restore from the versions backed up using the feature flag EnableAPIGroupVersions.
```

Author: Thanks, I'll get that changed.

## High-Level Design

During restore, the proposal is that Velero will read the directory names in the backup tarball file and see which API group versions were backed up from the source cluster. Velero will also look at the target cluster and determine which API group versions are supported. Given the two lists of source and target cluster supported API versions, a priority system will be used to determine which version will be backed up. The priority of versions in order are target cluster preferred version, source cluster preferred version, and the latest common supported version. Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.

## Detailed Design

There are four main objectives to achieve the above stated goal:

> 1. List the backed up API group versions.

Reviewer: I believe it will handle the backward-compatible on the source cluster, which means it will support both _without_ and _with_ enabling feature flag `EnableAPIGroupVersions` on the source cluster backed up tarball.

Or, should it only compatible with format version >= 1.1.0?
https://github.com/vmware-tanzu/velero/blob/4178d9d/pkg/apis/velero/v1/backup.go#L233-L235

Author: > it will support both without and with enabling feature flag

I'm not sure what you mean. Without the feature flag, only the `source preferred version` will be in the backup tarball. No other supported versions will be there so I can only use the source preferred version when the feature flag is not enabled.

As for format version, the new format version applies to the all backup tarballs, with or without the feature flag. But without the feature flag, only to the source preferred version will use format version 1.1.0 because no other version will be in the tarball.

Only if the feature flag is enabled will the other supported versions be available and these versions will meet format version 1.1.0.

Let me know if I misunderstood you.

Author: I'm not sure if this will help. Here is what's in the backed tarball files. This first screenshot shows the autoscaler versions that are backed up without feature flag enabled:

![Screen Shot 2020-11-03 at 8 09 30 AM](https://user-images.githubusercontent.com/17328443/98010982-2d81e500-1dac-11eb-866a-ceb8a4a5c857.png)

This second screenshot shows the autoscaler versions that are backed up with the feature flag enabled:
![Screen Shot 2020-11-03 at 8 09 09 AM](https://user-images.githubusercontent.com/17328443/98011037-3d012e00-1dac-11eb-857f-1c879e279e84.png)

You can see that in both cases, format version 1.1.0 is being used, but the difference is how many versions were backed up.

Reviewer: Both the screenshots if I remember correctly are commits format version `1.1.0`, it means that the source cluster and target cluster are running Velero after 1.4.

My question is will we support restores from the tarball with the legacy format version (Velero < 1.4)?
I guess yes because without enable the feature flag `EnableAPIGroupVersions`, the format version `1.1` and `1.1.0` almost look the same. Ref to https://github.com/vmware-tanzu/velero/blob/main/site/content/docs/main/output-file-format.md

Author: You bring up a really good point. Thank you for helping me to realize that what I proposed is not backward compatible with Velero versions earlier than 1.4.

If the backup tarball files do not follow `format version 1.1.0`, the code I have proposed will error when it shouldn't. It will try to set the chosen version to `v1-preferredversion`, but that version will not exist, and the restore will fail.

I'll update the design doc to check the directory names. If there is a directory with the substring `-preferredversion`, I know it is using `format version 1.1.0`. If not, I will let the program go on as usual, as if feature flag was not enabled.

Very good find, @jenting.

1. List the API group versions supported by the target cluster.

> 1. Use a priority system to determine which version to back up with the source preferred version as the default.

Reviewer: I was thought that step 3 would be `Use a priority system to determine which version to restore from with the target preferred version as the default.`

Because the priority system will determine which API version to be restored from by:
1. getting source API versions from the backup tarball
2. getting target API versions by discovery API
3. determine which API version to be restored from the backup tarball with order
   - target cluster preferred version if found in the backup tarball
   - source cluster preferred version if source cluster preferred version available in the target cluster
   - the latest common supported version for others

Author: The default will only take effect if all the three priorities fail. I chose `source preferred` so we can revert back to what the logic was like before the code changes with this PR. When our priority system fails, the failure will look the same as it would have failed before the feature flag was enabled.

Here are the sequence of events in case of failure:

1. Set default to `source cluster preferred version`.
2. Priority one fails. We cannot find a target preferred version that is supported by the source versions.
3. Priority two fails. We cannot find a source preferred version that is supported by the target versions.
4. Priority three fails. We cannot find a version that is common to the target and source.
5. Restore will use default and continue to report error per normal (backup version not supported by target cluster or something like that).

Reviewer: > The default will only take effect if all the three priorities fail. I chose `source preferred` so we can revert back to what the logic was like before the code changes with this PR. When our priority system fails, the failure will look the same as it would have failed before the feature flag was enabled.
> 
> Here are the sequence of events in case of failure:
> 
> 1. Set default to `source cluster preferred version`.
> 2. Priority one fails. We cannot find a target preferred version that is supported by the source versions.
> 3. Priority two fails. We cannot find a source preferred version that is supported by the target versions.
> 4. Priority three fails. We cannot find a version that is common to the target and source.
> 5. Restore will use default and continue to report error per normal (backup version not supported by target cluster or something like that).

Thanks for your clear explanation, a good idea if all the three priorities fail, it falls back to use `source cluster preferred version`.
I do not understand the sentence at my first glance `Use a priority system to determine which version to back up with the source preferred version as the default.`

Author: I admit that sentence is confusing. I'll split it up. Is this easier to understand:

`Use a priority system to determine which version to back up. The source preferred version will be the default if the priorities fail.`?

Reviewer: Use a priority system to determine which version to "back up", should it be "restore from" 🤔 ?

Author: haha oops. Another good catch @jenting. Should be 
> Use a priority system to determine which version to restore with the source preferred version as the default.

## Abstract

This document proposes a solution to select an API group version to restore from the versions backed up using the feature flag EnableAPIGroupVersions.

## Background

It is possible that between the time a backup has been made and a restore occurs that the target Kubernetes version has incremented more than one version. In such a case where multiple versions of Kubernetes were skipped, the preferred source cluster's API group versions for resources may no longer be supported by the target cluster. With PR #2373, all supported API group versions were backed up if the EnableAPIGroupVersions feature flag was set during Velero installation. The next step (outlined by this design proposal) will be to see if any of the backed up versions are supported in the target cluster and if so, choose one to restore for each backed up resource.

## Goals

- Restore resources using the chosen API group version.

## Non Goals

- Allowing users to customize the predetermined API group versions priorities.
- Modifying the compressed backup tarball files.
- Using plugins to restore a resource when the target supports non of the source cluster's API group versions.

## High-Level Design

> During restore, the proposal is that Velero will read the directory names in the backup tarball file and see which API group versions were backed up from the source cluster. Velero will also look at the target cluster and determine which API group versions are supported. Given the two lists of source and target cluster supported API versions, a priority system will be used to determine which version will be backed up. The priority of versions in order are target cluster preferred version, source cluster preferred version, and the latest common supported version. Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.

Reviewer: ```suggestion
During restore, the proposal is that Velero will read the directory names in the backup tarball file and see which API group versions were backed up from the source cluster. Velero will also look at the target cluster and determine which API group versions are supported. Given the two lists of source and target cluster supported API versions, a priority system will be used to determine which version will be restored. The priority of versions in order are target cluster preferred version, source cluster preferred version, and the latest common supported version. Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.
```

Author: 😆 I keep getting restore and backup confused. I'll get that into my next commit.

## Detailed Design

There are four main objectives to achieve the above stated goal:

1. Determine if the APIGroupVersionsFeatureFlag feature flag has been enabled.
1. List the backed up API group versions.
1. List the API group versions supported by the target cluster.
1. Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail.
1. Modify the paths to the backup files in the tarball in the resource restore process.

### Objective 1: Determine if the APIGroupVersionsFeatureFlag feature flag has been enabled

> For restore to be able to choose from multiple supported backed up versions, the feature flag must have been enabled during both the backup and restore processes. And the check to see if the feature flag is/was enabled must be done for both processes.

Reviewer: If I understand correctly, the design proposal _only_ supports the feature flag `APIGroupVersionsFeatureFlag` enabled on both the source and target cluster.
If yes, it'd be good to add this limitation to `Non Goals` chapter.

Author: Good call. Yes, it will be a good limitation to add in the Non Goals section.

The `chooseAPIVersionsToRestore` method can remain in the `restore` package, but for organizational purposes, it can be moved to a file called `prioritize_group_version.go`.

Inside the `chooseAPIVersionsToRestore` method, we can take advantage of the `archive` package's `Parser` type. A new parser object can be created using the construction function `NewParser(...)`. A method for the parser object will be created that has the signature `ParseGroupVersions(backupDir string) (map[string]metav1.APIGroup, error)`. The `ParseGroupVerisons(...)` method will loop through the `resources`, `resource.group`, and group version directories to populate a map called `sourceRGVersions`. 

The `sourceRGVersions` map's keys will be strings in the format `<resource>.<group>`, e.g. "horizontalpodautoscalers.autoscaling". The values will be APIGroup structs. The API Group struct can be imported from k8s.io/apimachinery/pkg/apis/meta/v1.

### Objective 3: List the API group versions supported by the target cluster

Still within the `chooseAPIVersionsToRestore` method, the target cluster's resource group versions can now be obtained.

```go
targetGroupVersions := ctx.discoveryHelper.APIGroups()
```

> ### Objective 4: Use a priority system to determine which version to back up, if any

Reviewer: ```suggestion
### Objective 4: Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail
```

Author: That makes sense. I'll add that change to my next commit.

# Restore API Group Version by Priority Level When EnableAPIGroupVersions Feature is Set

Status: Draft

## Abstract

This document proposes a solution to select an API group version to restore from the versions backed up using the feature flag EnableAPIGroupVersions.

## Background

> It is possible that between the time a backup has been made and a restore occurs that the target Kubernetes version has incremented more than one version. In such a case where multiple versions of Kubernetes were skipped, the preferred source cluster's API group versions for resources may no longer be supported by the target cluster. With PR #2373, all supported API group versions were backed up if the EnableAPIGroupVersions feature flag was set during Velero installation. The next step (outlined by this design proposal) will be to see if any of the backed up versions are supported in the target cluster and if so, choose one to restore for each backed up resource.

Author: Yup, will do.

## Goals

- Restore resources using the chosen API group version.
- Choose an API group to restore from backups given a priority system or a user-provided prioritization of versions.

## Non Goals

- Allow users to restore onto a cluster that is is running a Kubernetes version older than the source cluster. The changes proposed here only allow for skipping ahead to a newer Kubernetes version, not older.
- Allow restoring from backups created using Velero version 1.3 or older.
- Modifying the compressed backup tarball files.

> - Using plugins to restore a resource when the target supports non of the source cluster's API group versions.

Author: Good catch.

## High-Level Design

During restore, the proposal is that Velero will determine if the `APIGroupVersionsFeatureFlag` was enabled in both the source and target clusters (during backup and restore, respectively). Only if the feature flag was enabled in both clusters will the changes proposed here take effect.

The proposed code starts with creating three lists for each backed up resource. The three lists will be created by
  (1) reading the directory names in the backup tarball file and seeing which API group versions were backed up from the source cluster,
  (2) looking at the target cluster and determining which API group versions are supported, and
  (3) getting config maps on the cluster in order to get user-defined prioritization of versions.

>   The three lists will be used to create a map of chosen versions to restore. If there is a user-defined list of priority versions, the versions will be checked against the lists of target and source supported versions. The highest user-defined priority version that is/was supported by both target and source clusters will be the chosen version for that resource.

Author: Yes, if the user provides a value that doesn't exist in the source and/or target (the value must be in both), then the next paragraph will be executed. The next paragraph is to follow our pre-determined version prioritization (e.g. target preferred version is the highest priority). 

I'll add a sentence that if the user specified version is not supported by both target and source, it will be logged and the restore will continue with other priortizations.

>   The three lists will be used to create a map of chosen versions to restore. If there is a user-defined list of priority versions, the versions will be checked against the lists of target and source supported versions. The highest user-defined priority version that is/was supported by both target and source clusters will be the chosen version for that resource.

Reviewer 2: Thanks! I found it later in the document too, but I think a brief sentence here would be helpful.

## Detailed Design

There are six objectives to achieve the above stated goals:

1. Determine if the APIGroupVersionsFeatureFlag feature flag has been enabled.
1. List the backed up API group versions.
1. List the API group versions supported by the target cluster.
1. Get the user-defined version priorities.
1. Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail.
1. Modify the paths to the backup files in the tarball in the resource restore process.

### Objective 1: Determine if the APIGroupVersionsFeatureFlag feature flag has been enabled

For restore to be able to choose from multiple supported backed up versions, the feature flag must have been enabled during both the backup and restore processes. And the check to see if the feature flag is/was enabled must be done for both processes.

The reason for checking for the feature flag during restore is to ensure the user would like to restore a version that is among potentially multiple versions backed up with the feature flag enabled.

> The reason for checking to see if the feature flag was enabled during backup is to ensure the changes made by this proposed design is backward compatible. Only with Velero version 1.4 and forward was Format Version 1.0.0 used to structure the backup directories. Format Version 1.0.0 is required for the restore process proposed in this design doc to work. Before v1.4, the backed up files were in a directory structure that will not be recognized by the proposed code changes. Therefore, restore should not attempt to restore from multiple versions as they will not exist.

Author: Good eye--will change that.

Checking if the feature flag is enabled during restore is straightforward using `features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag)`.

> Checking if the feature flag was enabled during backup can be done indirectly by checking the directory structure and version directory names. For backups generated using Format Version 1.0.0 (required for the changes proposed here), there will always be a version directory with `-preferredversion` in the directory name. Only one resource will need to be checked to determine if the feature flag was enabled during backup.

Reviewer 2: I think this paragraph is a little inaccurate.

A backup can be format 1.1.0 without the feature flag enabled, and should have the `-preferredversion` directory even without the feature flag. 

At backup time currently, the feature flag controls whether all versions are downloaded. By default, Velero only downloads the preferred version and stores it twice; once under the `-preferredversion` directory, once under a version-less directory.

Therefore, I think this could be simplified to only operate on `Backup` objects with `Status.FormatVersion == "1.1.0"`.  [Here's the actual const value](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/backup.go#L58).

Reviewer: I had submitted a PR #2901 previously to fix the describe backup format version incorrectly (which was released at v1.5.0).
So, the only concern for me is if the source cluster uses Velero 1.4.0 to performs backup, then we can't use the format version to do the check.

Author: Thank you for the suggestion and another very good catch.

You're right that the paragraph is inaccurate. I will change it to say that we'll check for `Status.FormatVersion == "1.1.0"`. I realize from your observation that it is ok that during backup, the feature flag was not enabled. What's important are these two things:
(1) Feature flag is enabled during restore
(2) FormatVersion == 1.1.0

I'll be sure to change my wording to reflect those two conditions.

Reviewer 2: @jenting From conversations with Rafael, Velero won't be able to do this migration with backups made prior to v1.4.0 anyway. The feature flag and support for any API version besides the preferred version was only introduced in v1.4.0.

Also, see [line 21 of this document](https://github.com/vmware-tanzu/velero/pull/3050/files#diff-85b71029d5725d4b9e77b52b2276f38caa0086332e139cbe9f49b9b7abb7401cR21), which states that doing migration with Velero versions 1.3.x or older is a non-goal.

@codegold79 Thanks! The important thing about the feature flag during backup is whether or not more than 1 version is backed up, so that version negotiation can take place. Having a backup of Format 1.1.0 means you'll get the `-preferredversion` directory, but it doesn't guarantee you'll get multiple API versions - that only happens when the feature flag is on. So I think that may be useful to check for still, though it's _also_ possible that a given API only has 1 version, anyway...which makes that heuristic kind of unreliable.

Off the top of my head, I don't think we write the feature flags that were enabled anywhere in the backup data, but maybe that would be useful so we don't have to use these kinds of nondeterministic heuristics.

Reviewer 2: @jenting Sorry, I misunderstood your comment.

The code wouldn't have to use `velero describe`, it would use the `FormatVersion` field directly, which should still be reliable. The PR you referenced is only about the display on `velero describe`, not whether the `FormatVersion` field gets set correctly.

Reviewer: Thanks for the detailed explanation 🙏, I have no other concerns.

For restore to be able to choose from multiple supported backed up versions, the feature flag must have been enabled during the restore processes. Backup objects must also have [Status.FormatVersion == "1.1.0"](https://github.com/vmware-tanzu/velero/blob/a1e182e723a8c5f6d4175d8db2361233a94d2502/pkg/backup/backup.go#L58).

The reason for checking for the feature flag during restore is to ensure the user would like to restore a version that might not be the source cluster preferred version.

The reason for checking `Status.FormatVersion` is to ensure the changes made by this proposed design is backward compatible. Only with Velero version 1.4 and forward was Format Version 1.1.0 used to structure the backup directories. Format Version 1.1.0 is required for the restore process proposed in this design doc to work. Before v1.4, the backed up files were in a directory structure that will not be recognized by the proposed code changes. In this case, restore should not attempt to restore from multiple versions as they will not exist.

Checking if the feature flag is enabled during restore is straightforward using `features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag)`.

> Checking the `Status.FormatVersion` can be done indirectly by checking the directory structure and version directory names. For backups generated using Format Version 1.1.0, there will always be a version directory with `-preferredversion` in the directory name. Only one resource will need to be checked to determine if the feature flag was enabled during backup.

Reviewer 2: Can this paragraph be deleted, since on L57 you're looking directly at the backup object's `Status.FormatVersion` field? Or is this a double check to be sure?

Reviewer 2: To clarify - [the Backup object is included in a restore Request](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L81).

It is [also on the restoreContext](https://github.com/vmware-tanzu/velero/blob/main/pkg/restore/restore.go#L229).

So we have direct access to the Velero `Backup` CR that's being restored, not just the tarball.

Author: Oh, I think I misread something you wrote earlier. I do see `status.formatVersion` inside `velero-backup.json`. Would this be where I grabbed the format version from?

And if I can do that, I definitely won't need to be looking through the directories to look for `-preferredversion` in the names.

Reviewer 2: I think in this case, it would end up being from the restoreContext.

Something like `ctx.backup.Status.FormatVersion`.

You shouldn't have to parse the `velero-backup.json` yourself, as that's done for you. The JSON file will end up in the `ctx.backup` variable as a full Go struct.

Author: Nice! Thanks so much @nrb. I wrote my last question after your clarification. I'll fix the paragraphs to use the backup info in the `restoreContext`. That's much better.

Reviewer 2: The `Backup` comes from [this location](https://github.com/vmware-tanzu/velero/blob/6808acd92e30848056a21faf373af03ddb8a3b71/pkg/controller/restore_controller.go#L403) up in the restore controller, as an FYI. :)

So it's not directly from the JSON file, but rather the in-cluster representation. The JSON file is used when Velero syncs into brand new clusters, though.

> To help with the format version check, a new parser object can be created using the construction function `NewParser(...)` and we can take advantage of the `archive` package's `Parser` type. A method for the parser object will be created that has the signature `UsedFormatVersion(dir string) (bool, error)`. The method will loop through the version directories inside one of the resource.group directories and determine if a directory with `-preferredversion` was found.

Reviewer 2: Similar question - is this paragraph, and the `Parser` needed since you're using the `Status.FormatVersion` filed directly instead of inferring it based on directory structure?

Author: I'd be able to get rid of `UsedFormatVersion` function, but I'll still need `Parser` to read the version directories, e.g. `v1-preferredversion`, `v2`, etc.

A map will be created to store the user-defined priority versions. The map's keys will be strings in the format `<resource>.<group>`. The values will be APIGroup structs that will be imported from `k8s.io/apimachinery/pkg/apis/meta/v1`. Within the APIGroup structs will be versions in the order that the user provides in the config map. The PreferredVersion field in APIGroup struct will be left empty.

### Objective 5: Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail

Determining the priority will also be done in the `chooseAPIVersionsToRestore` method. Once a version is chosen, it will be stored in a new map of the form `map[string]ChosenGRVersion` where the key is the `<resource>.<group>` and the values are of the `ChosenGrpVersion` struct type (shown below). The map will be saved to the `restore.Context` object in a field called `chosenGrpVersToRestore`.

```go

> type ChosenGrpVersion struct {

Reviewer 2: I think this could be completely spelled out in the type name.

Author: No problem. I'll name the type to `ChosenGroupVersion`.

## High-Level Design

During restore, the proposal is that Velero will determine if the `APIGroupVersionsFeatureFlag` was enabled in both the source and target clusters (during backup and restore, respectively). Only if the feature flag was enabled in both clusters will the changes proposed here take effect.

The proposed code starts with creating three lists for each backed up resource. The three lists will be created by
  (1) reading the directory names in the backup tarball file and seeing which API group versions were backed up from the source cluster,
  (2) looking at the target cluster and determining which API group versions are supported, and
  (3) getting config maps on the cluster in order to get user-defined prioritization of versions.

  The three lists will be used to create a map of chosen versions to restore. If there is a user-defined list of priority versions, the versions will be checked against the lists of target and source supported versions. The highest user-defined priority version that is/was supported by both target and source clusters will be the chosen version for that resource.

>   Without a user-defined prioritization of versions, the following version priority will be followed: target cluster preferred version, source cluster preferred version, and a common supported version. Should there be multiple common supported versions, the one that will be chosen will be based on the [Kubernetes version priorities](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority%E2%80%8B).

Reviewer: ```suggestion
  Without a user-defined prioritization of versions, the following version priority will be followed: target cluster preferred version, source cluster preferred version, and a common supported version. Should there be multiple common supported versions, the one that will be chosen will be based on the [Kubernetes version priorities](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority).
```

Author: I will get that changed next commit.

- does the first source version in its APIGroups.Versions slice match any of the target supported versions? If so, it will be put into the `ChosenGRVersion` map.
- loop through the second target version and the rest. Inside that loop, loop through the second source version and the rest. The first version to match will populate the `ChosenGRVersion` values for the resource being looked at.
- if none of the previous checks produce a chosen version, the source preferred version will be the default and the restore process will continue.

Here is another way to list the priority versions described above:

- **Priority 0** ((User override). Users determine restore version priority using a config map
- **Priority 1**. Target preferred version can be used. This means one of two things:
  - (A) target preferred version == source preferred version OR
  - (B) target preferred version == source supported version
- **Priority 2**. Source preferred version can be used. This means
  - source preferred version == target supported version
- **Priority 3**. A common supported version can be used. This means
  - target supported version == source supported version

>   - if multiple support versions intersect, choose the version using the [Kubernetes’ version prioritization system](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority%E2%80%8B)

Reviewer: ```suggestion
  - if multiple support versions intersect, choose the version using the [Kubernetes’ version prioritization system](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority)
```

Author: I will get that changed next commit.
